File: 백준/Gold/2632. 피자판매/피자판매.py
import sys
input=sys.stdin.readline

want=int(input())
a,b=map(int,input().split())
A=[]
for i in range(a):
  A.append(int(input()))
B=[]
for i in range(b):
  B.append(int(input()))

totalA=[A[0]]
for i in range(1,a*2-1):
  totalA.append(totalA[-1]+A[i%a])

infoA={totalA[a-1]:1}
totalA=[0]+totalA
for i in range(1,a+1):
  for j in range(i,i+a-1):
    now=totalA[j]-totalA[i-1]
    if now>want:
      continue
    if now in infoA:
      infoA[now]+=1
    else:
      infoA[now]=1

totalB=[B[0]]
for i in range(1,b*2-1):
  totalB.append(totalB[-1]+B[i%b])

infoB={totalB[b-1]:1}
totalB=[0]+totalB
for i in range(1,b+1):
  for j in range(i,i+b-1):
    now=totalB[j]-totalB[i-1]
    if now>want:
      continue
    if now in infoB:
      infoB[now]+=1
    else:
      infoB[now]=1

answer=0
if want in infoA:
  answer+=infoA[want]
if want in infoB:
  answer+=infoB[want]

for Akey in infoA.keys():
  left=want-Akey
  if left in infoB:
    answer+=infoA[Akey]*infoB[left]

print(answer)

File: 백준/Gold/2632. 피자판매/피자판매.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const h=Number(input[iter++])
  const [m,n]=input[iter++].split(' ').map(item=>Number(item))
  let A=[]
  for(let i=0;i<m;i++){
    A.push(Number(input[iter++]))
  }
  A=[...A,...A]
  let B=[]
  for(let i=0;i<n;i++){
    B.push(Number(input[iter++]))
  }
  B=[...B,...B]

  const infoA={}
  let totalA=0
  for(let i=0;i<m;i++){
    let now=0
    for(let j=i;j<i+m;j++){
      now+=A[j]
      if (j===i+m-1) totalA=now
      if (now in infoA) infoA[now]+=1
      else infoA[now]=1
    }
  }
  infoA[totalA]=1

  const infoB={}
  let totalB=0
  for(let i=0;i<n;i++){
    let now=0
    for(let j=i;j<i+n;j++){
      now+=B[j]
      if (j===i+n-1) totalB=now
      if (now in infoB) infoB[now]+=1
      else infoB[now]=1
    }
  }
  infoB[totalB]=1

  let answer=0
  const keys=Object.keys(infoA)
  for(let i=0;i<keys.length;i++){
    const left=h-Number(keys[i])
    if (left<0) continue
    else if (left===0) {
      answer+=infoA[keys[i]]
      continue
    }
    else if (!(left in infoB)) continue
    answer+=infoA[keys[i]]*infoB[left]
  }
  if (h in infoB) answer+=infoB[h]

  console.log(answer)
  process.exit()
})
